### PR TITLE
Advanced Routing Policies

### PR DESCRIPTION
**Pull request by soerenmartius**
This PR adds some more advanced routing policies as described in the [docs](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/routing-policy.html).

## Changes
- A records `name` will be set to the zones name if it isn't set explicitly. Therefore a records name becomes _optional_.

## Policies Added:
- [weighted routing policies](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/routing-policy.html#routing-policy-weighted)
- [failover routing policy](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/routing-policy.html#routing-policy-failover)

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -39,10 +39,12 @@ resource "aws_route53_zone" "zone" {
 locals {
   records_expanded = [
     for record in var.records : merge({
-      record_id = join("-", compact([lower(record.type), lower(record.name)]))
-      ttl       = null
-      records   = null
-      alias     = null
+      record_id      = join("-", compact([lower(record.type), lower(record.name), try("${lower(record.set_identifier)}-${lower(record.weight)}", null)]))
```
Comment (soerenmartius) [inline main.tf]: Not sure yet if this is really the right approach to go. I will decide later once I applied all the different policies.

File: examples/failover/main.tf [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+# ---------------------------------------------------------------------------------------------------------------------
+# We configure two records with associated healthchecks. Route53 will route the traffic to the secondary record if
+# the healthcheck of the primary record reports an unhealthy status.
+#
+# For more information on DNS failover with Route53, please check the related documentation:
+# https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/routing-policy.html#routing-policy-failover
+# ---------------------------------------------------------------------------------------------------------------------
+
+resource "aws_route53_health_check" "primary" {
```
Comment (soerenmartius) [inline examples/failover/main.tf]: The tests are failing as expected. We need to address the issue with computed values to get this working @mariux 
Screen Shot 2020-02-18 at 16 37 49

File: main.tf [hunk 3]
```diff
@@ -39,10 +39,20 @@ resource "aws_route53_zone" "zone" {
 locals {
   records_expanded = [
     for record in var.records : merge({
-      record_id = join("-", compact([lower(record.type), lower(record.name)]))
-      ttl       = null
-      records   = null
-      alias     = null
+      record_id = join("-", compact([lower(record.type), lower(record.name),
+        try(
+          "${lower(record.set_identifier)}-${lower(record.weight)}",                                    # weighted routing policy
```
Comment (mariux) [inline main.tf]: doesn't identifier so the job as this is something you can set in a unique way... adding other values seems too much as any change would create a new and delete another resource... makes plans really weird.. as it is not easy to spot recreation then..

File: main.tf [hunk 4]
```diff
@@ -39,10 +39,20 @@ resource "aws_route53_zone" "zone" {
...
+          null
+        ),
+      ]))
+      ttl             = null
+      records         = null
+      alias           = null
+      set_identifier  = null
+      weight          = null
+      failover        = null
+      health_check_id = null
```
Comment (mariux) [inline main.tf]: merge will remove null values anyway.. so there is no point adding them here

Comment (soerenmartius) [reply main.tf]: Screen Shot 2020-02-18 at 17 40 24

Comment (soerenmartius) [reply main.tf]: wasn't that introduced by you ? :)

Comment (mariux) [reply main.tf]: ;) yes... found this out when debugging the provider issue with alias... so can be removed.. just forgot to clean up afterwards as it seems... my bad..

Comment (mariux) [reply main.tf]: learning curve for the win!

Comment (mariux) [reply main.tf]: i assume we also do this in other modules... needs fixing i guess

Comment (soerenmartius) [reply main.tf]: doesn't look like it removes the null values.. if i remove the lines, the actual implementation breaks.

Screen Shot 2020-02-18 at 17 55 01

File: examples/ns-and-soa-management/main.tf [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+      ]
+    },
+    {
+      type = "SOA"
+      ttl  = 900
+      records = [
+        "ns-250.awsdns-31.com. awsdns-hostmaster.amazon.com. 1 7200 900 1209600 86400"
+      ]
+    }
+  ]
```
Comment (mariux) [inline examples/ns-and-soa-management/main.tf]: those are not editable.. as defined by the DNS afaik.. so what is the purpose?

File: main.tf [hunk 6]
```diff
@@ -39,10 +39,24 @@ resource "aws_route53_zone" "zone" {
 locals {
   records_expanded = [
     for record in var.records : merge({
-      record_id = join("-", compact([lower(record.type), lower(record.name)]))
-      ttl       = null
-      records   = null
-      alias     = null
+      # if no name is set for the record, we use the zones name
+      record_id = join("-", compact([lower(record.type), try(lower(record.name), ""),
```
Comment (mariux) [inline main.tf]: `""` will be removed by compact anyway... so not needed.

File: main.tf [hunk 7]
```diff
@@ -39,10 +39,24 @@ resource "aws_route53_zone" "zone" {
...
+        try(
+          "${lower(record.set_identifier)}-${lower(record.weight)}",                                    # weighted routing policy
+          "${lower(record.set_identifier)}-${lower(record.failover)}-${lower(record.health_check_id)}", # failover
```
Comment (mariux) [inline main.tf]: also `set_identifier` will do the job i guess it's not about naming resources by all variables they can have but giving it a uniq name...see other comment

File: main.tf [hunk 8]
```diff
@@ -39,10 +39,24 @@ resource "aws_route53_zone" "zone" {
...
+
+      name            = ""
+      ttl             = null
+      records         = null
+      alias           = null
+      allow_overwrite = null
+      set_identifier  = null
+      weight          = null
+      failover        = null
+      health_check_id = null
```
Comment (mariux) [inline main.tf]: will be removed by merge anyway...

File: main.tf (hunk 4 above)
Comment (soerenmartius) [reply main.tf]: @mariux

File: main.tf (hunk 8 above)
Comment (soerenmartius) [reply main.tf]: are you sure? If I remove these lines it will break the current implementation ( see my other comments )

File: main.tf (hunk 6 above)
Comment (mariux) [reply main.tf]: my  bad... missed the try()

Comment (soerenmartius) [reply main.tf]: true, resolved in 548c28a

File: examples/ns-and-soa-management/main.tf (hunk 5 above)
Comment (soerenmartius) [reply examples/ns-and-soa-management/main.tf]: resolved in 30ae3f7

File: main.tf (hunk 8 above)
Comment (mariux) [reply main.tf]: terraform console shows that result.... that is weird..

**Comment by soerenmartius** [comment]
Hi @mariux ,

I've cleaned up all the examples, wrote documentation, comments, etc.

**Some thoughts I'd like to share:**
- I feel that accepting either a single zone name or a list of names, causes a lot of complexity and duplicated code. E.G. `records_by_name` and `records_by_zone_id` share almost the same code. Wouldn't it be better to just create two separate zones and define the records as a single locale?

- Having a separate file for records, weighted_records and failover_records leads to a lot of duplicated code.  In theory, we could at least merge the files `weighted-records.tf` and `failover-records.tf`. The main question remains: _Do we really need to define separate resources for the different record types?_

- I think we misuse try. From the docs:

> Warning: The try function is intended only for concise testing of the presence of and types of object attributes. Although it can technically accept any sort of expression, we recommend using it only with simple attribute references and type conversion functions as shown in the examples above. Overuse of try to suppress errors will lead to a configuration that is hard to understand and maintain.

I will try to come up with an alternative solution, e.G. the [can expression](https://www.terraform.io/docs/configuration/functions/can.html) might be helpful.

**Comment by mariux** [comment]
we should also stop using variables in examples... as they are not meant to be modules but examples..? what do you think?

**Comment by soerenmartius** [comment]
> we should also stop using variables in examples... as they are not meant to be modules but examples..? what do you think?

As discussed, we need the examples / tests to be configurable to  to be able to set variables dynamically with terratest and to test the outputs.

**Pull request merged**